### PR TITLE
web: add TSLint

### PR DESCRIPTION
**Pull request by KasparEtter**

### CODE REVIEW

File: frontends/web/package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "postinstall": "npx tcm src --camelCase dashes",
     "build": "preact build --no-prerender --service-worker=false",
     "dev": "preact watch",
-    "lint": "eslint --cache --ignore-path ../../.gitignore \"**/*.{js,jsx}\"",
+    "lint": "eslint --cache --ignore-path ../../.gitignore '**/*.{js,jsx}'; exit_code_eslint=$?; tslint -e 'src/**/*.css.d.ts' 'src/**/*.ts?(x)'; exit_code_tslint=$?; if [ $exit_code_eslint -gt 0 -o $exit_code_tslint -gt 0 ]; then exit 1; fi",
```
Comment (x1ddos) [inline frontends/web/package.json]: Not sure `exit_code_xxx=$?` is going to work on Windows? Unless this is never run on Windows. Just checking.

File: frontends/web/tslint.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
+    "defaultSeverity": "error",
+    "extends": [
+        // https://github.com/palantir/tslint/blob/master/src/configs/recommended.ts
```
Comment (x1ddos) [inline frontends/web/tslint.json]: No comments in json format are allowed. Is this valid?

File: frontends/web/package.json (hunk 1 above)
Comment (x1ddos) [reply frontends/web/package.json]: Also, I'd use `-ne` instead of `-gt` to be more future proof.

File: frontends/web/tslint.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    "rules": {
+        "arrow-parens": [true, "ban-single-arg-parens"],
+        "interface-name": [true, "never-prefix"],
+        "max-line-length": [false],
+        "member-ordering": [false],
+        "object-literal-sort-keys": [false],
+        "quotemark": [true, "single", "jsx-double", "avoid-escape"]
+    },
+    "rulesDirectory": []
+}
```
Comment (thisconnect) [inline frontends/web/tslint.json]: having comments in .json files still weirds me out but I can live with it if you insist

File: frontends/web/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (KasparEtter) [reply frontends/web/package.json]: I will. 🙂

File: frontends/web/src/components/confirm/Confirm.jsx [hunk 4]
```diff
@@ -25,6 +25,7 @@ export default class Confirm extends Component {
     }
 
     componentDidMount() {
+        // @ts-ignore (override changes the method signature)
```
Comment (x1ddos) [inline frontends/web/src/components/confirm/Confirm.jsx]: How about making `this.confirmOverride` conform to the signature of `window.confirm` instead?
Maybe in a follow up change or create a bug?

File: frontends/web/src/app.jsx [hunk 5]
```diff
@@ -137,10 +141,12 @@ export class App extends Component {
                         <Settings
                             path="/settings"
                             deviceIDs={deviceIDs} />
+                        {/* Use with TypeScript: {Route<{ deviceID: string }>({ path: '/manage-backups/:deviceID', component: ManageBackups })} */}
                         <ManageBackups
                             path="/manage-backups/:deviceID"
-                            showCreate={true}
-                            deviceIDs={deviceIDs} />
+                            // showCreate={true} // Does not exist!
+                            // deviceIDs={deviceIDs} // Does not exist!
```
Comment (x1ddos) [inline frontends/web/src/app.jsx]: Maybe just remove these? If you do it in a separate commit on top of the existing work, then it can still be traced back in git history, i.e. you'd see first commented out with "// does not exist" and then removed completely in a next commit.

File: frontends/web/src/components/language/language.jsx [hunk 6]
```diff
@@ -14,11 +14,11 @@
  * limitations under the License.
  */
 
-import { Component } from 'preact';
+import { Component, h } from 'preact';
 import { translate } from 'react-i18next';
 // import { Button } from '../forms';
 // import Dialog from '../dialog/dialog';
-// import style from './language.css';
+// import * as style from './language.css';
```
Comment (x1ddos) [inline frontends/web/src/components/language/language.jsx]: Remove these as well?

Comment (x1ddos) [reply frontends/web/src/components/language/language.jsx]: I mean, it doesn't feel like a `Dialog` should have anything to do with the language anyway, in the language module that is.

**Comment by x1ddos** [comment]
There's only one thing I noticed in the [CI output](https://travis-ci.org/digitalbitbox/bitbox-wallet-app/jobs/431115645):

```
$ jest --ci --no-color --coverage
PASS test/utils/equal.test.tsx
PASS test/components/message/message.test.tsx
Failed to collect coverage from /opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/src/utils/custom.d.ts
ERROR: Debug Failure. False expression: Output generation failed
STACK: Error: Debug Failure. False expression: Output generation failed
    at Object.transpileModule (/opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/typescript/lib/typescript.js:96861:18)
    at transpileViaTranspileModule (/opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/ts-jest/dist/transpiler.js:12:30)
    at Object.transpileTypescript (/opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/ts-jest/dist/transpiler.js:7:14)
    at Object.preprocess [as process] (/opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/ts-jest/dist/preprocess.js:31:40)
    at ScriptTransformer.transformSource (/opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/jest-runtime/build/script_transformer.js:314:35)
    at exports.default (/opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/jest-cli/build/generateEmptyCoverage.js:24:41)
    at Object.worker (/opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/jest-cli/build/reporters/coverage_worker.js:52:80)
    at /opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/jest-cli/build/reporters/coverage_reporter.js:257:45
    at Generator.next (<anonymous>)
    at step (/opt/go/src/github.com/digitalbitbox/bitbox-wallet-app/frontends/web/node_modules/jest-cli/build/reporters/coverage_reporter.js:75:30)
```

I guess it's nothing to worry about. Wouldn't expect `frontends/web/src/utils/custom.d.ts` to be included in the test coverage anyway.

File: frontends/web/src/components/confirm/Confirm.jsx (hunk 4 above)
Comment (KasparEtter) [reply frontends/web/src/components/confirm/Confirm.jsx]: Yes, I both mentioned it in our internal communication channel and already created an internal issue yesterday. I just wanted to keep the commit as focussed as possible/reasonable (and let other clean up their own mess).

Comment (x1ddos) [reply frontends/web/src/components/confirm/Confirm.jsx]: Yeah, I saw the comm but missed the bug. Just wanted to make sure. Thanks!

File: frontends/web/src/app.jsx (hunk 5 above)
Comment (KasparEtter) [reply frontends/web/src/app.jsx]: Yes, thought about this as well but especially the `showCreate` property that doesn't exist on the other route seems like a bug that might need proper fixing (and its own issue, which I still have to create). Plus, I wanted to refactor app.jsx to app.tsx anyway because TypeScript struggles with the implicit insertion of `deviceID` by the router. This is the reason why I decided to opt out completely of TS checking for the entire file: https://github.com/digitalbitbox/bitbox-wallet-app/pull/71/files/29014f71f7a2849bd07c8f3e5c05f1b41863e74c#diff-bc69595256b9d37e9f0a9653514577c7R17
I couldn't find out how to specify the generic type for TS through JSDoc comments (as I could do in other places) and just ignoring the following line with `@ts-ignore` didn't work for some reason… 🙁

Comment (KasparEtter) [reply frontends/web/src/app.jsx]: I just saw that `ManageBackups` has the `showCreate={true}` on its child and I will remove those lines in another PR then.
https://github.com/digitalbitbox/bitbox-wallet-app/blob/a85484ae01ed4dc8295915bbb3c5a311048fd4da/frontends/web/src/routes/device/manage-backups/manage-backups.jsx#L36-L45

File: frontends/web/src/components/language/language.jsx (hunk 6 above)
Comment (KasparEtter) [reply frontends/web/src/components/language/language.jsx]: What do you mean? I likely changed this with a global search and replace. The file describes the language switcher (and not a more general "language module"), which we only deactivated because we didn't want to keep the German translation up-to-date for now. In any case, this was again not the focus of my PR. 🙂

Comment (x1ddos) [reply frontends/web/src/components/language/language.jsx]: Sorry, that was out of scope indeed. Good that you ignored it :)
